packages/auth/lib/Users.js

```diff
@@ -40,9 +41,84 @@ const {
   FRONTEND_BASE_URL
 } = process.env
 
-const signIn = async (_email, context, pgdb, req, consents) => {
+// in order of preference
+const enabledFirstFactors = async (email, pgdb) => {
+  const userWithDevices = await pgdb.query(`
+    SELECT
+      u.*,
+      jsonb_agg(d.*) as devices
+    FROM
+      users u
+    LEFT JOIN
+      devices d
+      ON u.id = d."userId"
+    WHERE
+      u.email = :email
+    GROUP BY
+      u.id
+  `, {
+    email
+  })
+    .then(result => {
+      const user = result[0]
+      if (!user) {
+        return user
+      }
+      return {
+        ...user,
+        devices: user.devices
+          .filter(Boolean)
+      }
+    })
+
+  const { APP, EMAIL_TOKEN } = TokenTypes
+
+  const enabledTokenTypes = [EMAIL_TOKEN]
+  if (userWithDevices && userWithDevices.devices.length > 0) {
+    enabledTokenTypes.push(APP)
+  }
+
+  const { preferredFirstFactor } = userWithDevices || {}
+  return enabledTokenTypes
+    .sort((a, b) => {
+      if (a === b) {
+        return 0
+      }
+      if (a === preferredFirstFactor) { // preferred always wins
+        return -1
+      }
+      if (b === preferredFirstFactor) {
+        return 1
+      }
+      if (a !== EMAIL_TOKEN && b === EMAIL_TOKEN) { // EMAIL_TOKEN always last
+        return -1
+      }
+      if (a === EMAIL_TOKEN && b !== EMAIL_TOKEN) {
+        return 1
+      }
+      return 0 // undefined order for places [1...n-1]
+    })
+}
+
+const setPreferredFirstFactor = async (user, tokenType = null, pgdb) => {
+  const enabledTokenTypes = await enabledFirstFactors(user._raw.email, pgdb)
+  if (tokenType && enabledTokenTypes.indexOf(tokenType) === -1) {
+    throw new TokenTypeNotEnabledError({ tokenType })
+  }
+  return pgdb.public.users.updateAndGetOne(
+    {
+      id: user.id
+    }, {
+      preferredFirstFactor: tokenType
+    }
+  )
+}
+
+const signIn = async (_email, context, pgdb, req, consents, _tokenType) => {
   if (req.user) {
-    return { phrase: '' }
+    // TODO: How can we make it explizit that a user is logged in,
```

Comment:
Currently, when an already authenticated client calls `signIn`, the mutation just returns with this pseudo data. It's not to bad, but there must be a better/cleaner solution. Throwing is not ideal, because it would show an error on the client, which is totally confusing. Returning something else is not allowed by the graphql schema, what to do?

Reply:
Get it. It behaves as it did before apart from now additional `tokenType` that is returned.

It is in deed a bit odd, would not throw an error either. Not in this Pull Request but future might bring maybe a slightly different response signature, one with phrase and token and expiry date and one describing that you are logged in already. Sort of a state, maybe a whatever is returned when querying for `me`?